aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -710,6 +710,18 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_attention(
       query_, key, value, attn_mask_, dropout_p, is_causal);
 }
 
+inline void validate_sdpa_input(
+    const Tensor& query_,
+    const Tensor& key,
+    const Tensor& value,
+    const c10::optional<Tensor>& attn_mask_,
+    double dropout_p,
+    bool is_causal){
+    TORCH_CHECK(query_.dtype() == key.dtype() && query_.dtype() == value.dtype(), "Query, key, and value must have the same dtype");
+    TORCH_CHECK(query_.device() == key.device() && query_.device() == value.device(), "Query, key, and value must be on the same device");
+    TORCH_CHECK(query_.dim() >= 2 && key.dim() >= 2 && value.dim() >=2, "Query, key, and value must have at least 2 dimensions");
```

Comment:
Do we assert that k and v have the same dimension? 

In theory as long as q and k are broadcastable and than the product is broadcastable all q, k and v could have different number of dims. E.g.:
``` python
torch.Size(            [8, 128, 32])
torch.Size(      [32, 8, 128, 32])
torch.Size([64, 32, 8, 128, 64])
```

However in practice is this aligned with actual use cases of this attention mechanism

Reply:
I think the resolution is that we allow it but we have properly guarded the fused kernels so this will be all math fall back